Documentation

ffevotte opened the pull request:
The documentation effort made by @Vincent-Sivadon is still under way, but I'm creating this WIP PR so that we have a space to discuss issues related to it.

vsivadon committed:
README

vsivadon committed:
Benchmarking subfolder deleted (now in TiledFactorization)

vsivadon committed:
README updates

vsivadon committed:
README update

LaurentPlagne committed:
LP proposal

vsivadon committed:
Exemple

vsivadon committed:
Merge branch 'documentation' of https://github.com/maltezfaria/DataFlowTasks.jl into documentation

vsivadon committed:
BugFix (plot)

vsivadon committed:
Removed documentation build

vsivadon committed:
BugFix: Labels

vsivadon committed:
Updates

vsivadon committed:
Updates

vsivadon committed:
New clean README.md

vsivadon committed:
No Labels BugFix

vsivadon committed:
README update

ffevotte commented:
One issue that was brought up in earlier discussions comes from the use of a halo or zeros in the Gauss-Seidel example. IIUC, this is currently implemented by using a modified version of `PseudoTiledMatrix` that is aware of the existence of a halo, but in which only the "interior" of the matrix is tiled. This makes it possible to make the algorithm loops completely unaware of the existence of a halo. In particular, all tiles can assume that they have neighboring data in all directions.

OTOH, an other approach would consist in making the matrix itself unaware of the halo. The algorithm itself should therefore take care of iterating only on valid indices. In terms of documentation, this would have the added benefit that boundary tiles have less than 4 neighbors, so that the number of data dependencies is not always the same. It would be interesting to see whether this can be handled in a friendly way by DataFlowTasks and, if so, to use Gauss Seidel as an example to show how to.

A first idea that comes to mind to declare dependencies that don't always exist would be something along those lines:
```
@R (i>2 ? A[i-1,:] : nothing)
```
But I'm not sure whether this already works out of the box from a syntax viewpoint. Maybe it also requires implementing new methods like
```
memory_overlap(::Nothing, ::Any) = false
```

maltezfaria commented on README.md:
````diff
@@ -7,3 +7,194 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+n = 1000
+A = ones(n)
+
+@dspawn write!(@R(A), 1)          # 1
+@dspawn write!(@R(A[1:500]), 2)   # 2
+@dspawn write!(@R(A[501:n]), 3)   # 3
+@dspawn begin                     # 4
+  @RW A
+  readwrite!(A)
+end
+
+DataFlowTasks.sync()
+```
+
+This will generate the DAG (Directed Acyclic Graph) below that represents the dependencies between tasks. This means that the task 2 and 3 can be run in parallel. We see how it's the memory that matters here.
+
+![Basic](graph.png)
````
Are you sure that this is the `DAG` generated? I seems like tasks 2 and 3 operate on new data, and not on a view of `A`, so I would suspect that they are independent of everything else.

vsivadon replied:
It isn't indeed, using views does work. It's related to the issue mentioned yesterday. I forgot about the copying happenning. As François said a while ago, we should make it clear in the doc to avoid those problems.

maltezfaria commented:
> One issue that was brought up in earlier discussions comes from the use of a halo or zeros in the Gauss-Seidel example. IIUC, this is currently implemented by using a modified version of `PseudoTiledMatrix` that is aware of the existence of a halo, but in which only the "interior" of the matrix is tiled. This makes it possible to make the algorithm loops completely unaware of the existence of a halo. In particular, all tiles can assume that they have neighboring data in all directions.
> 
> OTOH, an other approach would consist in making the matrix itself unaware of the halo. The algorithm itself should therefore take care of iterating only on valid indices. In terms of documentation, this would have the added benefit that boundary tiles have less than 4 neighbors, so that the number of data dependencies is not always the same. It would be interesting to see whether this can be handled in a friendly way by DataFlowTasks and, if so, to use Gauss Seidel as an example to show how to.
> 
> A first idea that comes to mind to declare dependencies that don't always exist would be something along those lines:
> 
> ```
> @R (i>2 ? A[i-1,:] : nothing)
> ```
> 
> But I'm not sure whether this already works out of the box from a syntax viewpoint. Maybe it also requires implementing new methods like
> 
> ```
> memory_overlap(::Nothing, ::Any) = false
> ```

I am lost here, can you guys give me a reference to what you are calling the Gauss-Seidel algorithm? I can't reconcile what I have in mind, which is the iterative method, to this *five-point stencil* description that I found on the docs 😕

ffevotte commented:
> I am lost here, can you guys give me a reference to what you are calling the Gauss-Seidel algorithm? I can't reconcile what I have in mind, which is the iterative method, to this five-point stencil description that I found on the docs confused

Yes, I couldn't agree more. I should have mentioned that this is also a question I asked Vincent yesterday. @Vincent-Sivadon seems to have found this terminology in a paper on dataflow parallelism that I haven't seen yet...

vsivadon committed:
README executable

vsivadon committed:
Reset function

vsivadon committed:
README basic exemple changes

vsivadon committed:
PseudoTiledMatrix not in the README anymore

vsivadon committed:
PseudoTiledMatrix not in the Cholesky example anymore

vsivadon committed:
PseudoTiledMatrix not in the Seidel example anymore

ffevotte committed:
Simplified local documentation generation

ffevotte committed:
Fixed missing image path

ffevotte commented on README.md:
```diff
@@ -7,3 +7,178 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+readwrite!(X) = (X .+= 2)
+n = 1000
+A = ones(n)
+@dspawn readwrite!(@RW(A))              # 1
+@dspawn write!(@R(@view A[1:500]), 2)   # 2
+@dspawn write!(@R(@view A[501:n]), 3)   # 3
+@dspawn readwrite!(@RW(A))              # 4
```
Shouldn't this be:
```
@dspawn readwrite!(@RW(A))              # 1
@dspawn write!(@W(@view A[1:500]), 2)   # 2
@dspawn write!(@W(@view A[501:n]), 3)   # 3
@dspawn readwrite!(@RW(A))              # 4
```

ffevotte replied:
I would also enclose all these calls in a block, otherwise if people copy-paste them one at a time each task will be long finished when the following task is spawned.

ffevotte commented on README.md:
```diff
@@ -7,3 +7,178 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+function cholesky_dft!(A, ts)
+    m,n = size(A)
+    n%ts != 0 && error("Tilesize doesn't fit the matrix")
+    tn = round(Int, n/ts)
```
Under the constraint that `n%ts==0`, the computation of `tn` could be simplified to `tn = n÷ts`

ffevotte commented on README.md:
````diff
@@ -7,3 +7,178 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+
+```julia
+tilerange(ti, ts) = (ti-1)*ts+1:ti*ts
+function cholesky_dft!(A, ts)
````
maybe name this version `cholesky_tiled!`

ffevotte replied:
Actually, I was thinking of renaming the serial version `cholesky_tiled!`, and keeping the name `cholesky_dft!` for the parallel version. This way we could have both versions alongside one another, which makes it easier for potential users to compare them. What do you think?

vsivadon replied:
Oh i understand, it's clearer this way.

ffevotte commented on README.md:
```diff
@@ -7,3 +7,178 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+                mul!(view(A,rj,rk), adjoint(view(A,ri,rj)), view(A,ri,rk))
+            end
+        end
+    end
```
Couldn't all the views manipulations here be simplified if we defined once and for all something like:
```
Atile = [view(A, tilerange(ti, ts), tilerange(tj, ts)) for ti in 1:tn, tj in 1:tn]
```
Then, `view(A, rj, rk)` can be simplified to `Atile[tj,tk]`

ffevotte commented on README.md:
```diff
@@ -7,3 +7,178 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+            end
+        end
+    end
+    DataFlowTasks.sync()
+    # Construct the factorized object
+    return Cholesky(A,'U',zero(LinearAlgebra.BlasInt))
```
Broader question here: aren't these lines morally equivalent to the following?
```
result = @dspawn Cholesky(@R(A), 'U', zero(LinearAlgebra.BlasInt))
return fetch(result)  # this is where synchronization happens implicitly
```
If so, I would tend to favour this latter formulation, because I think it is more composable: we only ensure that all computations modifying `A` are done. If this was inserted in a larger algorithm, we could possibly have other parts of the computation involving `DataFlowTask` without interfering with each other.

In other words, I kinda feel like `DataFlowTasks.sync()` potentially has wider side-effects than we probably want, so its use should perhaps not be encouraged too much (unless it is unavoidable, of course)

maltezfaria replied:
@ffevotte I agree with you. I wanted to mention this earlier, but forgot: we should discourage the use of `DataFlowTasks.sync`. In fact, perhaps we should be even more drastic and remove it altogether from the public API. As François mentioned, there are often better ways to sync the output before returning which will allow for better composability. The code in `TiledFactorization` already uses and identical approach https://github.com/maltezfaria/TiledFactorization/blob/a8a4f8e9c1cd347a6eaef60b0f38477077121908/src/cholesky.jl#L43-L44

vsivadon replied:
I see, i just removed it at one point so it would make the profiling cleaner (because at the end there's juste one "Default" task).
I think we can keep it though, it's like the MPI.Barrier() function, not recommended but still useful sometimes, for debugging, easier developpement, safety etc...

ffevotte commented on README.md:
````diff
@@ -7,3 +7,178 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+DataFlowTasks.dagplot()
+```
+
+This will generate the DAG (Directed Acyclic Graph) above that represents the dependencies between tasks. This means that the task 2 and 3 can be run in parallel. We see how it's the memory that matters here.
````
I don't see any DAG image in the doc?

ffevotte commented on README.md:
```diff
@@ -7,3 +7,178 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+DFT.setscheduler!(DFT.JuliaScheduler(50))
+
+# Context
+n  = 4096
+ts = 512
+A = rand(n, n)
+A = (A + adjoint(A))/2
+A = A + n*I
+
+# Compilation
+cholesky_dft!(copy(A), ts)
```
We should probably add one or two lines of code here to check whether the computation is correct

ffevotte commented on README.md:
```diff
@@ -7,3 +7,178 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+cholesky_dft!(A ,ts)
+
+# Plot
+DFT.plot(categories=["chol", "ldiv", "mul"])
```
The labels don't seem to be in accordance to the image shown below.

Also, now that the dag is plotted separately, maybe we should add the line to generate it.

ffevotte replied:
And also perhaps the lines illustrating profiling should be put in a separate block in the section below.

ffevotte commented on README.md:
```diff
@@ -7,3 +7,178 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+
+We'll cover in details the usage and possibilities of the visualization in the documentation.
+
+Note that the visualization tools are not loaded by default, it requires a `Makie` backend and/or `GraphViz` loaded in the REPL. It's meant to be used in development, so it won't pollute the environment you want to use DFT in.
```
Yep! If/when #26 (or something like it) gets merged, we'll have to update the doc here accordingly.

ffevotte commented on README.md:
```diff
@@ -7,3 +7,178 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+
+We compare the performances achieved with this version of the Cholesky factorization with the MKL one, and we obtain the next figure. Here the blocks are of size (256, 256).
+
+![Perf](scalability_lfaria.png)
```
Has this image has been obtained with the simplified code above? If it comes from the complete high-performance implementation in `TiledFactorization`, we should say so here.

ffevotte reviewed (commented):
Thanks for this write-up!

Here are my comments on the brand-new README.

ffevotte committed:
Typos & other minor modifications in the text of README.md

ffevotte commented:
A global comment on this branch is: is there a reason why we need to have generated documentation files (such as `build/*.html`) checked in the git repo ?

vsivadon committed:
README update after reviews

vsivadon committed:
Merge branch 'documentation' of https://github.com/maltezfaria/DataFlowTasks.jl into documentation

ffevotte commented on README.md:
```diff
@@ -7,3 +7,216 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+readwrite!(X) = (X .+= 2)
+n = 1000
+A = ones(n)
+function ex()
+    @dspawn readwrite!(@RW(A))              # 1
+    @dspawn write!(@W(@view A[1:500]), 2)   # 2
+    @dspawn write!(@W(@view A[501:n]), 3)   # 3
+    @dspawn readwrite!(@RW(A))              # 4
+end
```
either explicitly run `ex()`, or replace it with a `let` block (or a `begin` block)

ffevotte commented on README.md:
````diff
@@ -7,3 +7,216 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+
+```julia
+tilerange(ti, ts) = (ti-1)*ts+1:ti*ts
+function cholesky_tiled!(A, ts)
+    m,n = size(A)
+    n%ts != 0 && error("Tilesize doesn't fit the matrix")
+    tn = n÷ts
+
+    for ti in 1:tn
+        ri = tilerange(ti, ts)
+
+        # Diagonal cholesky serial factorization (I)
+        cholesky!(view(A,ri,ri))
+
+        # Left blocks update (II)
+        L = adjoint(UpperTriangular(view(A,ri,ri)))
+        for tj in ti+1:tn
+            rj = tilerange(tj, ts)
+            ldiv!(L, view(A,ri,rj))
+        end
+
+        # Submatrix update (III)
+        for tj in ti+1:tn
+            for tk in tj:tn
+                rj = tilerange(tj, ts)  ;  rk = tilerange(tk, ts)
+                mul!(view(A,rj,rk), adjoint(view(A,ri,rj)), view(A,ri,rk))
+            end
+        end
+    end
+
+    # Construct the factorized object
+    return Cholesky(A,'U',zero(LinearAlgebra.BlasInt))
+end
````
I think this version could benefit from the rewrite with `Atile` as well.

ffevotte commented on README.md:
```diff
@@ -7,3 +7,216 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+                Ajk = Atile[tj,tk]
+                Aik = Atile[ti,tk]
+                Aji = adjoint(Atile[ti,tj])
+                @dspawn mul!(@RW(Ajk), @R(Aji), @R(Aik), -1, 1) label="schur ($tj,$tk)"
```
I think we determined that it was better to express the read dependency to `Aij` directly instead of `adjoint(Aij)`.

ffevotte commented on README.md:
````diff
@@ -7,3 +7,216 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+end
+```
+
+If you run this code, you'll see `memory_overlap` warning on the fact that there's no method to evaluate the memory overlaping between an `Adjoint` and a `SubArray`. So we have to implement this method :
+
+```julia
+using DataFlowTasks: memory_overlap
+memory_overlap(U::Adjoint,A) = memory_overlap(U.parent,A)
+memory_overlap(U,L::Adjoint) = memory_overlap(L,U)
+```
````
Please don't advise users to do this: it is type piracy

Well, technically it isn't here because this code defines a new `memory_overlap` function instead of adding methods to `DataFlowTasks.memory_overlap`, but this version shouldn't work either...

ffevotte commented on README.md:
````diff
@@ -7,3 +7,216 @@ Status](https://github.com/maltezfaria/DataFlowTasks.jl/workflows/CI/badge.svg)]
+
+## Profiling
+
+The next results are obtained with a slightly modified version of the function presented above. We used the `LoopVectorization` for the serial cholesky (I), `TriangularSolve`'s `ldiv!` and `Octavian`'s `matmul_serial!`.
+
+The package `TiledFactorization` contains this implementation, you can use it with the following lines :
+
+```julia
+import Pkg
+Pkg.add("https://github.com/maltezfaria/TiledFactorization.git")
+using TiledFactorization: cholesky!
+```
````
I think profiling should be illustrated on the example developed above. After all, we're not showing off with high-performance here, we are merely presenting a use-case.

IMO, only the performance plots should be coming from the `TiledFactorization` package.

vsivadon committed:
README update

ffevotte committed:
Automatically generate README.md using Literate.jl

This allows testing whether all code in the README is runnable.

Literate.jl also brings collateral benefits, such as the automatic generation of
Jupyter notebooks.

ffevotte commented:
Commit 9085534 introduces a way to automatically generate the `README.md` file from a `Literate.jl` julia script (located in `docs/readme/README.jl`). A collateral benefit is that the same contents are now also generated in the form of a Jupyter notebook (which can be seen directly from [github](https://github.com/maltezfaria/DataFlowTasks.jl/blob/documentation/docs/readme/README.ipynb) or from [nbviewer](https://nbviewer.org/github/maltezfaria/DataFlowTasks.jl/blob/documentation/docs/readme/README.ipynb)).

The new version of `README.md` has been generated this way, and we can now be sure that all code in it is runnable (which incidentally increases the number of test cases, although these are run during documentation generation and are therefore not accounted for in the coverage analysis).

I'll now go ahead and try to resolve all remaining issues with `README.md`, which should get us close to the final contents. Maybe a last review from y'all would be good though, in order to make sure we're on the same page in terms of how we want to present `DataFlowTasks` to new users.

More work is still needed to finalize the full Documenter docs, especially if we want to include all examples. But maybe we can reach pretty quickly a "presentable" state, where all features are documented. And from there we can add more tutorials/examples (maybe using `DemoCards.jl` or `Literate.jl` to more easily ensure they are runnable).

ffevotte committed:
README.md: final version?

ffevotte committed:
Removed generated files in `docs/build`

ffevotte commented on docs/readme/README.jl:
```diff
@@ -0,0 +1,247 @@
+# Let's look at a simple example:
+
+using DataFlowTasks
+DataFlowTasks.reset!() #hide
```
Unresolved questions regarding `reset`:
- do we want to keep it? (cf. #24)
- even if so, do we want to advertize its use?

OTOH: if we don't want to have/advertize it, make sure removing it doesn't affect the outcome of the examples shown here.

ffevotte replied:
Oh, and if we want to keep `reset()`, we'll have to add tests for it.

ffevotte commented on docs/readme/README.jl:
```diff
@@ -0,0 +1,247 @@
+# The code below shows how to use this `cholesky_tiled!` function, as well as how to profile the program and get information about how tasks were scheduled:
+
+## DataFlowTasks environnement setup
+DataFlowTasks.reset!()
+DataFlowTasks.enable_log()
```
Are these lines needed? Would `resetlogger!` be enough?

ffevotte commented on docs/readme/README.jl:
```diff
@@ -0,0 +1,247 @@
+F = cholesky_tiled!(copy(A), ts)                           #hide
+                                                           #hide
+## Check results                                           #hide
+err = norm(F.L*F.U-A,Inf)/max(norm(A),norm(F.L*F.U))       #hide
+@show err                                                  #hide
+@assert err < eps(Float64)                                 #hide
```
Naive question: why not something simpler, like
```
err = norm(F.L*F.U - A) / norm(A)
```
Also: I arbitrarily put the tolerance to machine precision, but this might not be the right thing to do.

ffevotte reviewed (commented):
I think the README is now in a state where it can be reviewed by anyone willing to.

I've commented on the few issues that still need to be discussed collectively IMO.

ffevotte committed:
Review of the Documenter-based documentation

Examples have been restructured, but are not ready yet.

ffevotte committed:
Literate-based infrastructure to generate examples

For now it is used only on Cholesky
